### PR TITLE
Issue 154

### PR DESCRIPTION
**Pull request by jollopre**
- Integrate pre-fetching plan and phase from 25e2669115561213d10e970d974460a20cc8b9c8
- after_save callback for answer model with tests. #154
- rake task to update completed plans. #154
- statistics#completed_plans action done. #154
- HTML view for completed plans and tweaks to JS file associated. #154

### CODE REVIEW

**Comment by jollopre** [comment]
Note this commit requires to run at ruby level:
```
rake db: migrate
rake bugfix:set_plan_complete
```
and at JS leve:
```
npm install
```

This PR does not include the "Download CSV" functionality @stephaniesimms and @sjDCC. I would create a new issue to address that separately.

File: app/models/phase.rb [hunk 1]
```diff
@@ -112,14 +112,12 @@ def self.deep_copy(phase)
...
-  # Returns the number of answered question for the phase. It is assumed that the plan_id passed
-  # has this phase instance.
-  def num_answered_questions(plan_id)
-    n = 0
-    self.sections.each do |s|
-      n+= s.num_answered_questions(plan_id)
+  # Returns the number of answered question for the phase.
+  def num_answered_questions(plan)
+    return 0 if plan.nil?
+    return sections.reduce(0) do |m, s|
```
Comment (briri) [inline app/models/phase.rb]: nice use of reduce

File: app/models/section.rb [hunk 2]
```diff
@@ -23,13 +23,16 @@ def to_s
...
-  # Returns the number of answered questions for a given plan id
-  def num_answered_questions(plan_id)
-    n = 0
-    self.questions.each do |question|
-      n += question.plan_answers(plan_id).select{|answer| answer.is_valid?}.count
+  # Returns the number of answered questions for a given plan
+  def num_answered_questions(plan)
+    return 0 if plan.nil?
+    questions_hash = questions.reduce({}){ |m, q| m[q.id] = q; m }
+    return plan.answers.includes({ question: :question_format }, :question_options).reduce(0) do |m, a|
```
Comment (briri) [inline app/models/section.rb]: nice refactor to load these other tables before calling out to is_valid

File: lib/assets/javascripts/views/usage/index.js [hunk 3]
```diff
@@ -45,35 +58,51 @@ $(() => {
...
-      console.log('error: %o', jqXHR);
+      },
     });
   };
+  // Sends an AJAX request to our two current endpoints that generate yearly data
+  // (e.g. users_joined_api_v0_statistics_path, completed_plans_api_v0_statistics_path )
+  // and draws a barChart when success response is found
+  const initialise = () => {
+    // Only fire AJAX requests if topicToURL object has keys, i.e. topics mapping to URLs
+    if (Object.keys(topicToURL).length > 0) {
```
Comment (briri) [inline lib/assets/javascripts/views/usage/index.js]: Thank you. Initialize was being fired on every page before and causing JS errors. This should fix that

**Pull request merged**